src/coreclr/tests/src/JIT/opt/Tailcall/EarlyFlowOptExample.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+
+// Test case showing that we can have "more complex"
+// IR after a tail call if we do early flow opts.
+
+interface IX
+{
+}
+
+class X : IX
+{
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    bool P1(object o) => false;
+
+    bool P0(object o)
+    {
+        bool b = false;
+        if (b)
+        {
+            return false;
+        }
+        return P1(o);
+    }
+
+    // F3 needs to be too big to inline without being marked noinline,
```

Comment:
This is one of those tests that seems like it could be easily not testing what we want it to test with a few changes/improvements to JIT capabilities and heuristics. And we wouldn't know (or be notified about) when that happens. I wonder if there would be some way to specify for a test/function "I expect this to tail call", or "I expect this to be inlined", for example, and if it doesn't happen, the test fails (e.g., JIT assert).